doc,bgpv1: Refresh BGP Control Plane document structure (#30345)
cilium-team/cilium: Documentation/network/bgp-control-plane.rst

.. _Border Gateway Protocol: https://datatracker.ietf.org/doc/html/rfc4271

Prerequisites
-------------

- Cilium must be configured with IPAM mode ``cluster-pool``, ``kubernetes``, or ``multi-pool``.
- If you are using the older MetalLB-based :ref:`bgp` feature, it must be disabled.

Installation
------------

.. tabs::

  .. group-tab:: Helm

        Cilium BGP Control Plane can be enabled with Helm flag ``bgpControlPlane.enabled``
        set as true.

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
                --namespace kube-system \\
                --reuse-values \\
                --set bgpControlPlane.enabled=true
            $ kubectl -n kube-system rollout restart ds/cilium

  .. group-tab:: Cilium CLI

        .. include:: ../installation/cli-download.rst

>         Cilium BGP Control Plane can be enabled with the below command

Reviewer: Can we change 'below' to 'following'? It helps language translators.

```suggestion
        Cilium BGP Control Plane can be enabled with the following command
```

assignment of the Router ID may be supported. Follow the `#30333
<https://github.com/cilium/cilium/issues/30333/>`_ for updates.

Validating Peering Status
^^^^^^^^^^^^^^^^^^^^^^^^^

Once the ``CiliumBGPPeeringPolicy`` is applied, you can check the BGP peering
status with the Cilium CLI with the following command:

.. code-block:: shell-session

   $ cilium bgp peers
   Node                              Local AS   Peer AS   Peer Address     Session State   Uptime   Family         Received   Advertised
   node0                             64512      64512     10.0.0.1         established     10s      ipv4/unicast   0          0
                                                                                                    ipv6/unicast   0          0

Node Annotations
----------------

A ``CiliumBGPPeeringPolicy`` can apply to multiple nodes. When a
``CiliumBGPPeeringPolicy`` applies to one or more nodes each node will
instantiate one or more BGP routers as defined in ``virtualRouters``. However,
there are times when fine-grained control over an instantiated virtual router's
configuration needs to take place. To accomplish this, a Kubernetes annotation

> is defined which applies to Kubernetes Node resources.

Reviewer 2: ```suggestion
configuration needs to take place. This can be accomplished by applying a Kubernetes annotation to Kubernetes Node resources.
```

Author: Thanks! Applied your suggestion.

A single annotation is used to specify a set of configuration attributes
to apply to a particular virtual router instantiated on a particular
host.

The syntax of the annotation is as follows:

::

       cilium.io/bgp-virtual-router.{asn}="key=value,..."

The ``{asn}`` portion should be replaced by the virtual router's local ASN you
wish to apply these configuration attributes to. Multiple option key/value
pairs can be specified by separating them with a comma. When duplicate
key/value are defined, the last one will be used.

Overriding Router ID
^^^^^^^^^^^^^^^^^^^^

When Cilium is running on an IPv4 single-stack or a dual-stack, the BGP Control
Plane uses the IPv4 address assigned to the node as the BGP Router ID. When
running in an IPv6 single-stack or when the administrator needs to manually

> define the instantiated Router ID, the ``router-id`` key can be used.

Reviewer 2: let's mention here why, and that it should be an IPv4 address

Author: Thanks I added some clarification.

.. code-block:: shell-session

   $ kubectl annotate node <node-name> cilium.io/bgp-virtual-router.{asn}="router-id=10.0.0.2"

Listening on the Local Port
^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, the BGP Control Plane instantiates each virtual router without a

> listening port. That means the BGP router only do connect to the peer, but

Reviewer 2: ```suggestion
listening port. That means the BGP router can only initiate connections to the configured peers, but
```

Author: Thanks! Applied your suggestion.

cannot accept incoming connections. This is the default behavior because the
BGP Control Plane is designed to be able to use in environments where the other

> BGP router such as ``Bird`` is running on the same node. When it is required to

Reviewer 2: ```suggestion
BGP Control Plane is designed to work also in environments where some other
BGP router (such as ``Bird``) is running on the same node. When it is required to
```

Author: Thanks! Applied your suggestion.

accept incoming connections, the ``local-port`` key can be used to specify the
listening port.

.. code-block:: shell-session

   $ kubectl annotate node <node-name> cilium.io/bgp-virtual-router.{asn}="local-port=179"

Advertising PodCIDRs
--------------------

> BGP Control Plane can advertise PodCIDRs of the nodes to the BGP peers. This

Reviewer 2: I would suggest to write more explicitly that each node advertises only local PodCIDRs - and the PodCIDRs of those nodes that are not running BGP are not advertised at all. I have seen that this wasn't clear for some users in the past.

Author: Thanks, I clarified them with sentences like:

```
With this configuration, the BGP speaker on each node advertises the PodCIDR prefixes assigned to the local node.
```

and

```
BGP Control Plane can advertise PodCIDR prefixes of the nodes selected by the CiliumBGPPeeringPolicy to the BGP peers.
```

Prerequisites
-------------

- Cilium must be configured with IPAM mode ``cluster-pool``, ``kubernetes``, or ``multi-pool``.
- If you are using the older MetalLB-based :ref:`bgp` feature, it must be disabled.

Installation
------------

.. tabs::

  .. group-tab:: Helm

        Cilium BGP Control Plane can be enabled with Helm flag ``bgpControlPlane.enabled``
        set as true.

        .. parsed-literal::

            $ helm upgrade cilium |CHART_RELEASE| \\
                --namespace kube-system \\
                --reuse-values \\
                --set bgpControlPlane.enabled=true
            $ kubectl -n kube-system rollout restart ds/cilium

  .. group-tab:: Cilium CLI

        .. include:: ../installation/cli-download.rst

        Cilium BGP Control Plane can be enabled with the below command

        .. parsed-literal::

            $ cilium install |CHART_VERSION| --set bgpControlPlane.enabled=true

IPv4/IPv6 single-stack and dual-stack setup are supported. Note that
the BGP Control Plane can only advertise the route of the address
family that the Cilium is configured to use. You cannot advertise IPv4

> route when the Cilium Agent is configured to use IPv6 only. Opposite is

Reviewer 3: ```suggestion
routes when the Cilium Agent is configured to use only IPv6 address family. Opposite is
```

Reviewer: In addition to these changes, do you also mind doing the following?

```suggestion
route when the Cilium Agent is configured to use IPv6 only. The opposite is
```

also true.

Configure Peering
-----------------

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   metadata:
     name: rack0
   spec:
     nodeSelector:
       matchLabels:
         rack: rack0
     virtualRouters:
     - localASN: 64512
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512

All BGP peering topology information is carried in a ``CiliumBGPPeeringPolicy``
CRD. ``CiliumBGPPeeringPolicy`` can be applied to one or more nodes based on

> its ``nodeSelector`` fields. Only single ``CiliumBGPPeeringPolicy`` can be

Reviewer: Couple of grammar suggestions. The CRD only has a single `nodeSelector` in the spec, correct?

```suggestion
CRD. A ``CiliumBGPPeeringPolicy`` can be applied to one or more nodes based on
its ``nodeSelector`` field. Only a single ``CiliumBGPPeeringPolicy`` can be
```

sessions until only one policy matches the node.

.. warning::

   Applying another policy over an existing one will cause the BGP session to
   be cleared and causes immediate connectivity disruption. It is strongly
   recommended to test the policy in a staging environment before applying it
   to production.

Each ``CiliumBGPPeeringPolicy`` defines one or more ``virtualRouters``. The
virtual router defines a BGP router instance which is uniquely identified by
its ``localASN``. Each virtual router can have multiple ``neighbors`` defined.
The neighbor defines a BGP neighbor uniquely identified by its ``peerAddress``
and ``peerASN``. When ``localASN`` and ``peerASN`` are the same, iBGP peering
is used. When ``localASN`` and ``peerASN`` are different, eBGP peering is used.

Specifying Router ID (IPv6 single-stack only)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When Cilium is running on an IPv4 or a dual-stack, the BGP Router ID is
automatically derived from the IPv4 address assigned to the node. When Cilium
is running on an IPv6 single-stack cluster, the BGP Router ID must be
configured manually. This can be done by setting the annotation on the

> Kubernetes Node resource.

Reviewer: ```suggestion
Kubernetes Node resource:
```

.. code-block:: shell-session

   $ kubectl annotate node <node-name> cilium.io/bgp-virtual-router.64512="router-id=10.0.0.2"

Currently, you must set the annotation for each node. In the future, automatic
assignment of the Router ID may be supported. Follow the `#30333

> <https://github.com/cilium/cilium/issues/30333/>`_ for updates.

Reviewer: In this context, you're talking about applying an annotation to a Node resource, therefore I think it makes sense to capitalize Node here. Also, small grammar suggestion:

```suggestion
Currently, you must set the annotation for each Node. In the future, automatic
assignment of the Router ID may be supported. Follow `#30333
<https://github.com/cilium/cilium/issues/30333/>`_ for updates.
```

   node0                             64512      64512     10.0.0.1         established     10s      ipv4/unicast   0          0
                                                                                                    ipv6/unicast   0          0

Node Annotations
----------------

A ``CiliumBGPPeeringPolicy`` can apply to multiple nodes. When a
``CiliumBGPPeeringPolicy`` applies to one or more nodes each node will
instantiate one or more BGP routers as defined in ``virtualRouters``. However,
there are times when fine-grained control over an instantiated virtual router's
configuration needs to take place. This can be accomplished by applying a
Kubernetes annotation to Kubernetes Node resources.

A single annotation is used to specify a set of configuration attributes
to apply to a particular virtual router instantiated on a particular
host.

The syntax of the annotation is as follows:

::

       cilium.io/bgp-virtual-router.{asn}="key=value,..."

The ``{asn}`` portion should be replaced by the virtual router's local ASN you
wish to apply these configuration attributes to. Multiple option key/value
pairs can be specified by separating them with a comma. When duplicate

> key/value are defined, the last one will be used.

Reviewer: Small wording suggestion:

```suggestion
pairs can be specified by separating them with a comma. When duplicate
keys are defined with different values, the last key's value will be used.
```

> Overriding Router ID

Reviewer: Is there a difference between this section and the section above titled 'Specifying Router ID (IPv6 single-stack only)'?

Author: For me, it's slightly different. They are talking about the same configuration knob, but 'Specifying Router ID (IPv6 single-stack only)' describes the necessary step to make peering work, but this section talks about the optional settings (and it's not specific to IPv6, we can use this config for IPv4 or Dual-stack setup).

Reviewer: Ah ok that makes sense, thanks!

^^^^^^^^^^^^^^^^^^^^

When Cilium is running on an IPv4 single-stack or a dual-stack, the BGP Control
Plane can use the IPv4 address assigned to the node as the BGP Router ID
because Router ID is 32bit long, and we can rely on the uniqueness of the IPv4
address to make Router ID unique which is not the case for IPv6. Thus, when
running in an IPv6 single-stack, or when the auto assignment of the Router ID
is not desired, the administrator needs to manually define it. This can be
accomplished by setting the ``router-id`` key in the annotation.

.. code-block:: shell-session

   $ kubectl annotate node <node-name> cilium.io/bgp-virtual-router.{asn}="router-id=10.0.0.2"

Listening on the Local Port
^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, the BGP Control Plane instantiates each virtual router without a
listening port. That means the BGP router can only initiate connections to the
configured peers, but cannot accept incoming connections. This is the default
behavior because the BGP Control Plane is designed to work also in environments

> where some other BGP router (such as ``Bird``) is running on the same node.

Reviewer: Couple wording suggestions:

```suggestion
listening port. This means the BGP router can only initiate connections to the
configured peers, but cannot accept incoming connections. This is the default
behavior because the BGP Control Plane is designed to function in environments
where another BGP router (such as ``Bird``) is running on the same node.
```

=============================== ===================
Selector                        Field
------------------------------- -------------------
io.cilium.podippool.namespace   ``.meta.namespace``
io.cilium.podippool.name        ``.meta.name``
=============================== ===================

For additional details regarding CiliumPodIPPools, see the :ref:`ipam_crd_multi_pool` section.

Advertising Service Virtual IPs
-------------------------------

Type Load Balancer Services
^^^^^^^^^^^^^^^^^^^^^^^^^^^

BGP Control Plane can advertise the ingress IPs
(``status.loadBalancer.ingress[*].ip``) of a Service of type ``LoadBalancer``
to the BGP peers. This allows the BGP peers to reach the Service directly from
outside the cluster. When your upstream router supports Equal Cost Multi Path
(ECMP), you can use this feature to load balance traffic to the Service across
multiple nodes by advertising the same ingress IPs from multiple nodes.

You must first allocate ingress IPs to advertise them. By default, Kubernetes
doesn't provide a way to assign ingress IPs to a Service. Cluster administrator
is responsible for preparing a controller that assigns ingress IPs. Cilium

> supports it with :ref:`Load Balancer IPAM <lb_ipam>` feature.

Reviewer: Wording suggestion:

```suggestion
You must first allocate ingress IPs to advertise them. By default, Kubernetes
doesn't provide a way to assign ingress IPs to a Service. The cluster administrator
is responsible for preparing a controller that assigns ingress IPs. Cilium
supports assigning ingress IPs with the :ref:`Load Balancer IPAM <lb_ipam>` feature.
```

           app: foo
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512

This advertises the ingress IPs of all Services matching the ``.serviceSelector``.

If you wish to announce ALL services within the cluster, a ``NotIn`` match expression
with a dummy key and value can be used like:

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   spec:
     nodeSelector:
       matchLabels:
         rack: rack0
     virtualRouters:
     - localASN: 64512
       serviceSelector:
          matchExpressions:
             - {key: somekey, operator: NotIn, values: ['never-used-value']}
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512

There are a few special purpose selector fields which don't match on labels but
instead on other metadata like ``.meta.name`` or ``.meta.namespace``.

=============================== ===================
Selector                        Field
------------------------------- -------------------
io.kubernetes.service.namespace ``.meta.namespace``
io.kubernetes.service.name      ``.meta.name``
=============================== ===================

Load Balancer Class

> ~~~~~~~~~~~~~~~~~~~

Reviewer: I think this header may need to match the one for "Type Load Balancer Services"? This is supposed to fall under "Advertising Service Virtual IPs" correct?

Author: Load Balancer Class is specific to the Type Load Balancer service, so I think it should be in the current place.

https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class

Cilium supports the `loadBalancerClass
<https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`__.
When the load balancer class is set to ``io.cilium/bgp-control-plane`` or unspecified,
Cilium will announce the ingress IPs of the Service. Otherwise, Cilium will not announce
the ingress IPs of the Service.

externalTrafficPolicy

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer: I think this header may need to match the one for "Type Load Balancer Services"? This is supposed to fall under "Advertising Service Virtual IPs" correct?

Author: We only handle externalTrafficPolicy for Load Balancer service as of today (we don't have any other Service VIP advertisement, we may have ClusterIP advertisement as well in the future). Thus I think it's reasonable to be here.

Reviewer: Sounds good, thanks for clarifying.

there's no local endpoint.

Validating Advertised Routes
----------------------------

Get all IPv4 unicast routes available:

.. code-block:: shell-session

   $ cilium bgp routes available ipv4 unicast
   Node                              VRouter   Prefix        NextHop   Age    Attrs
   node0                             64512     10.1.0.0/24   0.0.0.0   17m42s [{Origin: i} {Nexthop: 0.0.0.0}]

Get all IPv4 unicast routes available for a specific vrouter:

.. code-block:: shell-session

   $ cilium bgp routes available ipv4 unicast vrouter 64512
   Node                              VRouter   Prefix        NextHop   Age    Attrs
   node0                             64512     10.1.0.0/24   0.0.0.0   17m42s [{Origin: i} {Nexthop: 0.0.0.0}]

Get IPv4 unicast routes advertised to a specific peer:

.. code-block:: shell-session

   $ cilium bgp routes advertised ipv4 unicast peer 10.0.0.1
   Node                              VRouter   Prefix        NextHop   Age    Attrs
   node0                             64512     10.1.0.0/24   10.0.0.2  17m42s [{Origin: i} {AsPath: } {Nexthop: 10.0.0.2} {LocalPref: 100}]

> Neighbor Options

Reviewer 3: Can you also add `families` section under neighbor options.

Author: Let me handle this in the separate PR 🙏

field.

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   spec:
     nodeSelector:
       matchLabels:
         rack: rack0
     virtualRouters:
     - localASN: 64512
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512
         peerPort: 1179 # <-- specify the peer port

Timers
^^^^^^

BGP Control Plane supports modifying the following BGP timer parameters. For
more detailed description for each timer parameters, please refer to `RFC4271
<https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`__.

================= ============================ ==========
Name              Field                        Default
----------------- ---------------------------- ----------
ConnectRetryTimer ``connectRetryTimeSeconds``  120
HoldTimer         ``holdTimeSeconds``          90
KeepaliveTimer    ``keepAliveTimeSeconds``     30
================= ============================ ==========

In datacenter networks which Kubernetes clusters are deployed in general, it is
recommended to set the ``HoldTimer`` and ``KeepaliveTimer`` to a lower value

> for faster failure detection if possible. For example, you can set the minimum

Reviewer: Wording suggestion:

```suggestion
In datacenter networks which Kubernetes clusters are deployed, it is generally
recommended to set the ``HoldTimer`` and ``KeepaliveTimer`` to a lower value
for faster possible failure detection. For example, you can set the minimum
```

   level=error msg="Failed to fetch secret \"secretname\": not found (will continue with empty password)" component=manager.fetchPeerPassword subsys=bgp-control-plane

Graceful Restart
^^^^^^^^^^^^^^^^

> The Cilium BGP control plane can be configured to act as a graceful restart

Reviewer: Would you mind capitalizing `control plane` here to match the rest of the changes that you made?

Author: Sure! Fixed.
